flink-state-backends/flink-statebackend-forst/src/main/java/org/apache/flink/state/forst/fs/cache/CachedDataInputStream.java

```diff
@@ -102,62 +102,81 @@ private FSDataInputStream getStream() throws IOException {
         if (isFlinkThread()) {
             cacheEntry.touch();
         }
-        FSDataInputStream stream = tryGetCacheStream();
-        if (stream != null) {
-            fileBasedCache.incHitCounter();
-            return stream;
-        }
-
-        if (streamStatus == StreamStatus.CACHED_CLOSED
-                || streamStatus == StreamStatus.CACHED_CLOSING) {
-            if (streamStatus == StreamStatus.CACHED_CLOSING) {
-                try {
-                    semaphore.acquire(1);
-                } catch (InterruptedException e) {
-                    throw new RuntimeException(e);
-                }
-                originalStream.seek(position);
-                position = -1;
-                LOG.trace(
-                        "Stream {} status from {} to {}",
-                        cacheEntry.cachePath,
-                        streamStatus,
-                        StreamStatus.CACHED_CLOSED);
-                streamStatus = StreamStatus.CACHED_CLOSED;
-            }
-            // try reopen
-            tryReopen();
-            stream = tryGetCacheStream();
+        // Repeat until get a stream
+        while (true) {
+            // Firstly, we try to get cache stream
+            FSDataInputStream stream = tryGetCacheStream();
             if (stream != null) {
                 fileBasedCache.incHitCounter();
                 return stream;
             }
-            fileBasedCache.incMissCounter();
-            return originalStream;
-        } else if (streamStatus == StreamStatus.ORIGINAL) {
-            fileBasedCache.incMissCounter();
-            return originalStream;
-        } else {
-            if (streamStatus == StreamStatus.CACHED_OPEN) {
-                stream = tryGetCacheStream();
+
+            // No cache stream, so is it closing?
+            if (streamStatus == StreamStatus.CACHED_CLOSED
```

Comment:
nit:
we could refactor the code to remove the if with the or in so it is 
if (streamStatus == StreamStatus.CACHED_CLOSING) {
...
}
if (streamStatus == StreamStatus.CACHED_CLOSED) {
}